scheduler/cloudformation/cloudformation.go

```diff
@@ -1,5 +1,554 @@
+	RunTask(*ecs.RunTaskInput) (*ecs.RunTaskOutput, error)
+	RegisterTaskDefinition(*ecs.RegisterTaskDefinitionInput) (*ecs.RegisterTaskDefinitionOutput, error)
+	StopTask(*ecs.StopTaskInput) (*ecs.StopTaskOutput, error)
+	UpdateService(*ecs.UpdateServiceInput) (*ecs.UpdateServiceOutput, error)
+}
+
+// s3Client duck types the s3.S3 interface that we use.
+type s3Client interface {
+	PutObject(*s3.PutObjectInput) (*s3.PutObjectOutput, error)
+}
+
+// Template represents something that can generate a stack body. Conveniently
+// the same interface as text/template.Template.
+type Template interface {
+	Execute(wr io.Writer, data interface{}) error
+}
+
+// Templates should add metadata to the ECS resources with the following
+// structure.
+type serviceMetadata struct {
+	// This is the name of the process that this ECS service is for.
+	Name string `json:"name"`
+}
+
+// Scheduler implements the scheduler.Scheduler interface using CloudFormation
+// to provision resources.
+type Scheduler struct {
+	// Template is a text/template that will be executed using the
+	// twelvefactor.Manifest as data. This template should return a valid
+	// CloudFormation JSON manifest.
+	Template Template
+
+	// The ECS cluster to run tasks in.
+	Cluster string
+
+	// If true, wait for stack updates and creates to complete.
+	Wait bool
+
+	// The name of the bucket to store templates in.
+	Bucket string
+
+	// CloudFormation client for creating stacks.
+	cloudformation cloudformationClient
+
+	// ECS client for performing ECS API calls.
+	ecs ecsClient
+
+	// S3 client to upload templates to s3.
+	s3 s3Client
+
+	db *sql.DB
+}
+
+// NewScheduler returns a new Scheduler instance.
+func NewScheduler(db *sql.DB, config client.ConfigProvider) *Scheduler {
+	return &Scheduler{
+		cloudformation: cloudformation.New(config),
+		ecs:            ecs.New(config),
+		s3:             s3.New(config),
+		db:             db,
+	}
+}
+
+// Submit creates (or updates) the CloudFormation stack for the app.
+func (s *Scheduler) Submit(ctx context.Context, app *scheduler.App) error {
+	tx, err := s.db.Begin()
+	if err != nil {
+		return err
+	}
+
+	err = s.submit(ctx, tx, app)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	return tx.Commit()
+}
+
+// Submit creates (or updates) the CloudFormation stack for the app.
+func (s *Scheduler) submit(ctx context.Context, tx *sql.Tx, app *scheduler.App) error {
+	stackName, err := s.stackName(app.ID)
+	if err == errNoStack {
+		stackName = app.Name
+		if _, err := tx.Exec(`INSERT INTO stacks (app_id, stack_name) VALUES ($1, $2)`, app.ID, stackName); err != nil {
+			return err
+		}
+	} else if err != nil {
+		return err
+	}
+
+	buf := new(bytes.Buffer)
+	if err := s.Template.Execute(buf, app); err != nil {
+		return err
+	}
+
+	key := fmt.Sprintf("%x", sha1.Sum(buf.Bytes()))
+
+	_, err = s.s3.PutObject(&s3.PutObjectInput{
+		Bucket:      aws.String(s.Bucket),
+		Key:         aws.String(fmt.Sprintf("/%s", key)),
```

Comment:
since we have app id, i would include that in this key.

i could see it being useful to view different stack "versions" for a specific app.